File: modules/widgets/lib/rin/widgets/TransformInfo.cpp
#include "rin/widgets/TransformInfo.hpp"

#include "rin/widgets/WidgetContainerSlot.hpp"
#include "rin/widgets/WidgetSurface.hpp"
#include "rin/widgets/Widget.hpp"

TransformInfo::TransformInfo(const Matrix3<float>& inTransform, const Vec2<float>& inSize,int inDepth)
    {
        transform = inTransform;
        size = inSize;
        depth = inDepth;
    }

    TransformInfo::TransformInfo(const WidgetSurface* inSurface) : TransformInfo(Matrix3<float>{1.0f},inSurface->GetDrawSize().Cast<float>())
    {
        
    }

    TransformInfo::TransformInfo(const Shared<WidgetSurface>& inSurface) : TransformInfo(Matrix3<float>{1.0f},inSurface->GetDrawSize().Cast<float>())
    {
        
    }

    TransformInfo::TransformInfo(const Shared<Widget>& widget, bool absolute) : TransformInfo(absolute ? widget->ComputeAbsoluteTransform() : widget->ComputeRelativeTransform(),widget->GetDrawSize())
    {
        
    }

    TransformInfo TransformInfo::AccountFor(const Shared<WidgetContainerSlot>& slot) const
    {
        return AccountFor(slot->GetWidget());
    }

    TransformInfo TransformInfo::AccountFor(const Shared<Widget>& widget) const
    {
        return AccountFor(widget.get());
    }

    TransformInfo TransformInfo::AccountFor(const Widget* widget) const
    {
        auto newTransform = transform * widget->ComputeRelativeTransform();
        return TransformInfo{newTransform,widget->GetDrawSize(),depth + 1};
    }
    
    bool TransformInfo::IsPointWithin(const Vec2<float>& point) const
    {
        auto tl = Vec2<float>(0.0f);
        auto br =  size;
        auto tr = Vec2<float>(br.x, tl.y);
        auto bl = Vec2<float>(tl.x, br.y);

        tl =  transform * tl;
        br = transform * br;
        tr = transform * tr;
        bl = transform * bl;

        auto p1AABB = Vec2<float>(
            std::min(std::min(tl.x,tr.x),std::min(bl.x,br.x)),
            std::min(std::min(tl.y,tr.y),std::min(bl.y,br.y))
        );
        auto p2AABB = Vec2<float>(
            std::max(std::max(tl.x,tr.x),std::max(bl.x,br.x)),
            std::max(std::max(tl.y,tr.y),std::max(bl.y,br.y))
        );

        WidgetRect aabb(p1AABB,p2AABB - p1AABB);
        
        if(!aabb.Contains(point)) return false;


        auto top = tr - tl;
        auto right = br - tr;
        auto bottom = bl - br;
        auto left = tl - bl;
        auto pTop = point - tl;
        auto pRight = point - tr;
        auto pBottom = point - br;
        auto pLeft = point - bl;
        
        auto a = top.Acos(pTop);
        auto b = right.Cross(pRight);
        auto c = bottom.Cross(pBottom);
        auto d = left.Cross(pLeft);

        if (a >= 0)
        {
            return b >= 0 && c >= 0 && d >= 0;
        }
        else
        {
            return b < 0 && c < 0 && d < 0;
        }
    }


    WidgetRect TransformInfo::ComputeAxisAlignedBoundingRect() const
    {
        auto tl = Vec2<float>(0.0f);
        auto br =  size;
        auto tr = Vec2<float>(br.x, tl.y);
        auto bl = Vec2<float>(tl.x, br.y);

        tl =  transform * tl;
        br = transform * br;
        tr = transform * tr;
        bl = transform * bl;

        auto p1AABB = Vec2<float>(
            std::min(std::min(tl.x,tr.x),std::min(bl.x,br.x)),
            std::min(std::min(tl.y,tr.y),std::min(bl.y,br.y))
        );
        auto p2AABB = Vec2<float>(
            std::max(std::max(tl.x,tr.x),std::max(bl.x,br.x)),
            std::max(std::max(tl.y,tr.y),std::max(bl.y,br.y))
        );

        return {p1AABB,p2AABB - p1AABB};
    }

File: aerox.Runtime.Widgets/Graphics/TransformInfo.cs
using aerox.Runtime.Math;

namespace aerox.Runtime.Widgets.Graphics;

public struct ClippingRect
{
    public Matrix3 Matrix;
    public Size2d Size;
}

public class TransformInfo : ICloneable<TransformInfo>
{
    public Matrix3 Transform;
    public Size2d Size;
    public int Depth;
    
    /// <summary>
    /// True if the target of this draw info will be occluded (tests are done using an AABB approach
    /// </summary>
    public bool Occluded = false;

    public TransformInfo(Matrix3 transform,Size2d size,int depth)
    {
        Transform = transform;
        Size = size;
        Depth = depth;
    }
    
    public TransformInfo(Surface surface)
    {
        Transform = Matrix3.Identity;
        Size = surface.GetDrawSize();
        Depth = 0;
    }

    public Rect ToRect() => ComputeAxisAlignedBoundingRect(this);

    public TransformInfo Clone()
    {
        return new TransformInfo(Transform,Size,Depth);
    }

    //public static Vector2<float>[] ClippingRectToPoints()


    public static Rect ComputeAxisAlignedBoundingRect(TransformInfo target)
    {
        var tl = new Vector2<float>(0.0f);
        var br = tl + target.Size;
        var tr = new Vector2<float>(br.X, tl.Y);
        var bl = new Vector2<float>(tl.X, br.Y);

        tl = tl.ApplyTransformation(target.Transform);
        br = br.ApplyTransformation(target.Transform);
        tr = tr.ApplyTransformation(target.Transform);
        bl = bl.ApplyTransformation(target.Transform);

        var p1AABB = new Vector2<float>(
            System.Math.Min(
                System.Math.Min(tl.X, tr.X),
                System.Math.Min(bl.X, br.X)
            ),
            System.Math.Min(
                System.Math.Min(tl.Y, tr.Y),
                System.Math.Min(bl.Y, br.Y)
            )
        );
        var p2AABB = new Vector2<float>(
            System.Math.Max(
                System.Math.Max(tl.X, tr.X),
                System.Math.Max(bl.X, br.X)
            ),
            System.Math.Max(
                System.Math.Max(tl.Y, tr.Y),
                System.Math.Max(bl.Y, br.Y)
            )
        );

        return new Rect
        {
            Offset = p1AABB,
            Size = p2AABB - p1AABB
        };
    }

    public bool IntersectsWith(TransformInfo target)
    {
        var targetAARect = ComputeAxisAlignedBoundingRect(target);
        var myAARect = ComputeAxisAlignedBoundingRect(this);
        return targetAARect.IntersectsWith(myAARect);
    }

    public bool PointWithin(Vector2<float> point)
    {
        var tl = new Vector2<float>(0.0f);
        var br = tl + Size;
        var tr = new Vector2<float>(br.X, tl.Y);
        var bl = new Vector2<float>(tl.X, br.Y);

        tl = tl.ApplyTransformation(Transform);
        br = br.ApplyTransformation(Transform);
        tr = tr.ApplyTransformation(Transform);
        bl = bl.ApplyTransformation(Transform);

        var p1AABB = new Vector2<float>(
            System.Math.Min(
                System.Math.Min(tl.X, tr.X),
                System.Math.Min(bl.X, br.X)
            ),
            System.Math.Min(
                System.Math.Min(tl.Y, tr.Y),
                System.Math.Min(bl.Y, br.Y)
            )
        );
        var p2AABB = new Vector2<float>(
            System.Math.Max(
                System.Math.Max(tl.X, tr.X),
                System.Math.Max(bl.X, br.X)
            ),
            System.Math.Max(
                System.Math.Max(tl.Y, tr.Y),
                System.Math.Max(bl.Y, br.Y)
            )
        );

        // Perform AABB test first
        if (!point.Within(new Rect
            {
                Offset = p1AABB,
                Size = p2AABB - p1AABB
            })) return false;

        var top = tr - tl;
        var right = br - tr;
        var bottom = bl - br;
        var left = tl - bl;
        var pTop = point - tl;
        var pRight = point - tr;
        var pBottom = point - br;
        var pLeft = point - bl;
        
        var a = top.Acos(pTop);
        var b = right.Cross(pRight);
        var c = bottom.Cross(pBottom);
        var d = left.Cross(pLeft);

        if (a >= 0)
        {
            return b >= 0 && c >= 0 && d >= 0;
        }
        else
        {
            return b < 0 && c < 0 && d < 0;
        }
    }
}